compiler+runtime/src/cpp/jank/runtime/module/loader.cpp

```diff
@@ -338,47 +339,101 @@ namespace jank::runtime::module
       return err(fmt::format("unable to find module: {}", module));
     }
 
-    string_result<void> res{ err(fmt::format("no sources for registered module: {}", module)) };
+    auto const with_module_type{ [&](module_type const type) {
+      loader::find_result res{ entry->second, type };
+      return res;
+    } };
 
     //fmt::println("loading nested module {}", module);
 
-    /* When we're compiling, we always load from source. Otherwise, we
-     * load from source if we specifically chose to do so. */
-    bool const compiling{ truthy(rt_ctx.compile_files_var->deref()) };
-    if(compiling || ori == origin::source)
+    if(ori == origin::source)
     {
       if(entry->second.jank.is_some())
       {
-        res = load_jank(entry->second.jank.unwrap());
+        return with_module_type(module_type::jank);
       }
       else if(entry->second.cljc.is_some())
       {
-        res = load_cljc(entry->second.cljc.unwrap());
+        return with_module_type(module_type::cljc);
       }
     }
     else
     {
-      /* TODO: origin must be latest, so we need to find that. Even if there
-       * is a binary, we need to only choose it if it's as new, or newer, than
-       * the source. */
-      if(entry->second.o.is_some())
+      if(entry->second.o.is_some()
+         && (entry->second.jank.is_some() || entry->second.cljc.is_some()))
       {
-        res = load_o(module, entry->second.o.unwrap());
+        auto const o_file_path{ native_transient_string{ entry->second.o.unwrap().path } };
+
+        native_transient_string source_path{};
+        module_type module_type{};
+
+        if(entry->second.jank.is_some())
+        {
+          source_path = entry->second.jank.unwrap().path;
+          module_type = module_type::jank;
+        }
+        else
+        {
+          source_path = entry->second.cljc.unwrap().path;
+          module_type = module_type::cljc;
+        }
+
+        if(boost::filesystem::last_write_time(o_file_path)
+           >= boost::filesystem::last_write_time(source_path))
```

Comment:
Do these throw if the file no longer exists? It's possible the file has been deleted since we first learned about it. We need to be sure not to throw here. Check the docs and protect these calls accordingly. If a file has gone missing, or we can't access it, we should use the other. If both are bonked, then we can return an error.